File: orion-tenant-svc-go/internal/service/tenant_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"regexp"

	"orion/tenant-svc-go/internal/models"
	"orion/tenant-svc-go/internal/repository"

	"github.com/google/uuid"
	"go.uber.org/zap"
)

var errInvalidName = errors.New("tenant name can only contain letters, numbers, hyphens and underscores")
var nameRegex = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

type TenantService struct {
	repo *repository.TenantRepository
	log  *zap.Logger
}

func NewTenantService(repo *repository.TenantRepository, log *zap.Logger) *TenantService {
	return &TenantService{repo: repo, log: log}
}

// GetTenant retrieves a tenant by ID.
func (s *TenantService) GetTenant(ctx context.Context, id string) (*models.Tenant, error) {
	t, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if t == nil {
		return nil, fmt.Errorf("tenant not found: %s", id)
	}
	return t, nil
}

// GetTenantByName retrieves a tenant by name.
func (s *TenantService) GetTenantByName(ctx context.Context, name string) (*models.Tenant, error) {
	return s.repo.FindByName(ctx, name)
}

// ListTenants returns paginated tenants.
func (s *TenantService) ListTenants(ctx context.Context, page, limit int, status string) ([]models.Tenant, int, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 {
		limit = 20
	}
	offset := (page - 1) * limit

	tenants, err := s.repo.List(ctx, &status, limit, offset)
	if err != nil {
		return nil, 0, err
	}

	var total int
	if status != "" {
		total, err = s.repo.Count(ctx, &status)
	} else {
		total, err = s.repo.Count(ctx, nil)
	}
	if err != nil {
		return nil, 0, err
	}
	return tenants, total, nil
}

// CreateTenant creates a new tenant.
func (s *TenantService) CreateTenant(ctx context.Context, name string, displayName string) (*models.Tenant, error) {
	name = toLowerCase(name)
	if name == "" {
		return nil, fmt.Errorf("tenant name is required")
	}
	if !nameRegex.MatchString(name) {
		return nil, errInvalidName
	}

	exists, err := s.repo.ExistsByName(ctx, name)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, fmt.Errorf("tenant name already exists: %s", name)
	}

	var dn *string
	if displayName != "" {
		dn = &displayName
	}

	t := &models.Tenant{
		ID:          uuid.New().String(),
		Name:        name,
		DisplayName: dn,
		Status:      "active",
		Settings:    make(map[string]any),
	}
	if err := s.repo.Create(ctx, t); err != nil {
		return nil, err
	}
	return t, nil
}

// UpdateTenant updates a tenant.
func (s *TenantService) UpdateTenant(ctx context.Context, id string, updates map[string]any) (*models.Tenant, error) {
	existing, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if existing == nil {
		return nil, fmt.Errorf("tenant not found: %s", id)
	}

	// Check duplicate name
	if name, ok := updates["name"]; ok {
		newName, ok := name.(string)
		if ok && newName != existing.Name {
			exists, _ := s.repo.ExistsByName(ctx, newName)
			if exists {
				return nil, fmt.Errorf("tenant name already exists: %s", newName)
			}
		}
	}

	return s.repo.Update(ctx, id, updates)
}

// DeleteTenant performs a soft delete.
func (s *TenantService) DeleteTenant(ctx context.Context, id string) error {
	existing, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return err
	}
	if existing == nil {
		return fmt.Errorf("tenant not found: %s", id)
	}
	return s.repo.SoftDelete(ctx, id)
}

// toLowerCase converts a string to lowercase.
func toLowerCase(s string) string {
	result := make([]byte, len(s))
	for i := 0; i < len(s); i++ {
		b := s[i]
		if b >= 'A' && b <= 'Z' {
			b = b + 32
		}
		result[i] = b
	}
	return string(result)
}

File: legacy/orion-platform-service-ts/src/services/tenant/TenantService.ts
/**
 * TenantService - Business logic layer for Tenant operations
 * 
 * Handles business rules, validation, and orchestration
 */

import { 
  TenantRepository, 
  Tenant, 
  CreateTenantInput, 
  UpdateTenantInput 
} from './TenantRepository';

export interface ListTenantsOptions {
  page?: number;
  limit?: number;
  status?: string;
}

export interface PaginatedResult<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export class TenantServiceError extends Error {
  constructor(message: string, public code: string) {
    super(message);
    this.name = 'TenantServiceError';
  }
}

export class TenantService {
  private repository: TenantRepository;

  constructor(repository: TenantRepository) {
    this.repository = repository;
  }

  /**
   * Get tenant by ID
   */
  async getTenant(id: string): Promise<Tenant> {
    const tenant = await this.repository.findById(id);
    
    if (!tenant) {
      throw new TenantServiceError(`Tenant not found: ${id}`, 'TENANT_NOT_FOUND');
    }
    
    return tenant;
  }

  /**
   * Get tenant by name
   */
  async getTenantByName(name: string): Promise<Tenant | null> {
    return this.repository.findByName(name);
  }

  /**
   * List all tenants with pagination
   */
  async listTenants(options: ListTenantsOptions = {}): Promise<PaginatedResult<Tenant>> {
    const { page = 1, limit = 20, status } = options;
    const offset = (page - 1) * limit;

    const [tenants, total] = await Promise.all([
      this.repository.findAll({ status, limit, offset }),
      this.repository.count(status),
    ]);

    return {
      data: tenants,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    };
  }

  /**
   * Create a new tenant
   */
  async createTenant(input: CreateTenantInput): Promise<Tenant> {
    // Validate input
    if (!input.name || input.name.trim().length === 0) {
      throw new TenantServiceError('Tenant name is required', 'INVALID_INPUT');
    }

    // Check for duplicate name
    const exists = await this.repository.existsByName(input.name);
    if (exists) {
      throw new TenantServiceError('Tenant name already exists', 'DUPLICATE_NAME');
    }

    // Validate name format (alphanumeric, hyphens, underscores)
    if (!/^[a-zA-Z0-9_-]+$/.test(input.name)) {
      throw new TenantServiceError(
        'Tenant name can only contain letters, numbers, hyphens and underscores',
        'INVALID_NAME_FORMAT'
      );
    }

    return this.repository.create({
      ...input,
      name: input.name.toLowerCase().trim(),
      display_name: input.display_name?.trim(),
    });
  }

  /**
   * Update an existing tenant
   */
  async updateTenant(id: string, input: UpdateTenantInput): Promise<Tenant> {
    // Check if tenant exists
    const existing = await this.repository.findById(id);
    if (!existing) {
      throw new TenantServiceError(`Tenant not found: ${id}`, 'TENANT_NOT_FOUND');
    }

    // Check for duplicate name if name is being changed
    if (input.name) {
      const exists = await this.repository.existsByName(input.name);
      if (exists && existing.name !== input.name) {
        throw new TenantServiceError('Tenant name already exists', 'DUPLICATE_NAME');
      }
    }

    const updated = await this.repository.update(id, input);
    
    if (!updated) {
      throw new TenantServiceError(`Failed to update tenant: ${id}`, 'UPDATE_FAILED');
    }
    
    return updated;
  }

  /**
   * Delete a tenant (soft delete)
   */
  async deleteTenant(id: string): Promise<boolean> {
    // Check if tenant exists
    const existing = await this.repository.findById(id);
    if (!existing) {
      throw new TenantServiceError(`Tenant not found: ${id}`, 'TENANT_NOT_FOUND');
    }

    return this.repository.delete(id);
  }

  /**
   * Permanently delete a tenant (use with caution)
   */
  async hardDeleteTenant(id: string): Promise<boolean> {
    const existing = await this.repository.findById(id);
    if (!existing) {
      throw new TenantServiceError(`Tenant not found: ${id}`, 'TENANT_NOT_FOUND');
    }

    return this.repository.hardDelete(id);
  }
}